pkg/services/ngalert/sender/sender.go

```diff
@@ -171,29 +176,104 @@ func buildNotifierConfig(cfg *ngmodels.AdminConfiguration) (*config.Config, erro
+			sanitizedLabelName = sanitizedLabelName + fmt.Sprintf("_%.3x", md5.Sum([]byte(k)))
+			s.logger.Warn("alert contains duplicate label/annotation name after sanitization, appending unique suffix", "name", k, "new_name", sanitizedLabelName, "err", err)
+		}
 
-	// Prometheus does not allow spaces in labels or annotations while Grafana does, we need to make sure we
-	// remove them before sending the alerts.
-	for k, v := range alert.Alert.Labels {
-		ls = append(ls, labels.Label{Name: removeSpaces(k), Value: v})
+		set[sanitizedLabelName] = struct{}{}
+		ls = append(ls, labels.Label{Name: sanitizedLabelName, Value: lbls[k]})
 	}
 
-	for k, v := range alert.Annotations {
-		a = append(a, labels.Label{Name: removeSpaces(k), Value: v})
+	return ls
+}
+
+// sanitizeLabelName will fix a given label name so that it is compatible with prometheus alertmanager character restrictions.
+// Prometheus alertmanager requires labels to match ^[a-zA-Z_][a-zA-Z0-9_]*$.
+// Characters with an ASCII code < 127 will be replaced with an underscore (_), characters with ASCII code >= 127 will be replaced by their hex representation.
+// For backwards compatibility, whitespace will be removed instead of replaced with an underscore.
+func (s *ExternalAlertmanager) sanitizeLabelName(name string) (string, error) {
+	if len(name) == 0 {
+		return "", errors.New("label name cannot be empty")
 	}
 
-	return &notifier.Alert{
-		Labels:       ls,
-		Annotations:  a,
-		StartsAt:     time.Time(alert.StartsAt),
-		EndsAt:       time.Time(alert.EndsAt),
-		GeneratorURL: alert.Alert.GeneratorURL.String(),
+	if isValidLabelName(name) {
+		return name, nil
+	}
+
+	s.logger.Warn("alert sending to external Alertmanager(s) contains label/annotation name with invalid characters", "name", name)
+
+	// Remove spaces. We do this instead of replacing with underscore for backwards compatibility as this existed before the rest of this function.
+	sanitized := strings.Join(strings.Fields(name), "")
+
+	// Replace other invalid characters.
+	var buf strings.Builder
+	for i, b := range sanitized {
+		// From alertmanager LabelName.IsValid().
+		if (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || b == '_' || (b >= '0' && b <= '9' && i > 0) {
+			buf.WriteRune(b)
+		} else {
+			if b <= unicode.MaxASCII {
+				buf.WriteRune('_')
+			} else {
+				_, _ = fmt.Fprintf(&buf, "%x", b)
+			}
+		}
+	}
+
+	if buf.Len() == 0 {
+		return "", fmt.Errorf("label name only contains invalid chars")
 	}
+
+	return buf.String(), nil
 }
 
-func removeSpaces(labelName string) string {
-	return strings.Join(strings.Fields(labelName), "")
+// isValidLabelName is true iff the label name matches the pattern of ^[a-zA-Z_][a-zA-Z0-9_]*$. From alertmanager LabelName.IsValid().
+func isValidLabelName(ln string) bool {
+	if len(ln) == 0 {
+		return false
+	}
+	for i, b := range ln {
+		if !((b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || b == '_' || (b >= '0' && b <= '9' && i > 0)) {
```

Comment:
there is a similar expression above. I think it is complex enough to be extracted to a function. What do you think?